File: android/src/main/kotlin/io/github/kokemus/cross_bluetooth_api/commands/Command.kt
package io.github.kokemus.cross_bluetooth_api.commands

enum class CommandId(val method: String) {
    REQUEST_DEVICE("requestDevice"),
    CONNECT("connect"),
    DISCONNECT("disconnect"),
    GET_PRIMARY_SERVICE("getPrimaryService"),
    GET_CHARACTERISTIC("getCharacteristic"),
    READ_VALUE("readValue"),
    WRITE_VALUE_WITHOUT_RESPONSE("writeValueWithoutResponse"),
    WRITE_VALUE_WITH_RESPONSE("writeValueWithResponse"),
    START_NOTIFICATIONS("startNotifications"),
    STOP_NOTIFICATIONS("stopNotifications");

    companion object {
        fun fromMethod(method: String): CommandId? = entries.find { it.method == method }
    }
}

interface Command {
    suspend fun execute()
}

File: ios/cross_bluetooth_api/Sources/cross_bluetooth_api/Commands/Command.swift
import Foundation
import Flutter

enum CommandId: String, Codable {
    case requestDevice
    case connect
    case disconnect
    case getPrimaryService
    case getCharacteristic
    case readValue
    case writeValueWithoutResponse
    case writeValueWithResponse
    case startNotifications
    case stopNotifications
}

protocol Command {
    func execute() async
}

public class BaseCommand: NSObject, Command {
    let id: CommandId
    let arguments: [String: AnyObject]
    let pendingResult: FlutterResult
    
    init(
        id: CommandId, 
        arguments: [String: AnyObject]?,
        pendingResult: @escaping FlutterResult,
    ) {
        self.id = id
        self.arguments = arguments ?? [:]
        self.pendingResult = pendingResult
    }

    func execute() async {
        pendingResult(FlutterError(code: "NotImplemented", message: "Command \(id.rawValue) is not implemented.", details: nil))
    }
}
